openedx/core/lib/cache_utils.py

```diff
@@ -151,6 +153,57 @@ def __get__(self, obj, objtype):
         return functools.partial(self.__call__, obj)
 
 
+class CacheInvalidationManager:
+    """
+    This class provides a decorator for simple functions, which can handle invalidation.
+
+    To use, instantiate with a namespace or django model class:
+    `manager = CacheInvalidationManager(model=User)`
+
+    Then use it as a decorator on functions with no arguments
+    `@manager
+    def get_system_user():
+         ...
+    `
+    When the User model is saved or deleted, all cache keys used by
+    the decorator will be cleared.
+    """
+
+    def __init__(self, namespace=None, model=None, cache_time=86400):
+        if model:
+            post_save.connect(self.invalidate, sender=model)
+            post_delete.connect(self.invalidate, sender=model)
+            namespace = str(model)
+        self.namespace = namespace
+        self.cache_time = cache_time
+        self.keys = set()
+
+    # pylint: disable=unused-argument
+    def invalidate(self, **kwargs):
+        """
+        Invalidate all keys tracked by the manager.
+        """
+        for key in self.keys:
+            TieredCache.delete_all_tiers(key)
```

Comment:
Could you please change this to use the request cache instead of the tiered cached? Setting things in memcached across processes doesn't buy that much performance, but it could cause weird bugs because memcached lives outside the transaction lifetime of the database. So for instance, if we do a save on a model within a transaction, and that transaction later gets rolled back, using the tiered cache like this would have already done a set in memcached with that model value. Then that bad/rolled-back value would be alive in memcached for another process that runs.

Reply:
Also, that way we don't have to worry about what happens when we serialize into memcache an older version of a model that just had a column added to it, etc.

Reply:
I made this invalidation class in order to cache the `XBlockConfiguration` queryset, which rarely changes (and only through the deployment process or django admin, right?). I don't think there's much win by storing it in the request, but it seems appropriate to keep it in memcached.

Reply:
Ah. I was more worried about the long term if this started being used for things like enrollments. I agree this is safe to do for something like config models.

But if that's the use case, then I think we were hitting at the same problem from different sides. :-(

https://github.com/edx/django-config-models/pull/55

Reply:
Ah, that will be a good change. When will it land? Can I help?

But it won't solve this particular problem -- the xblock config is accessed like `XBlockConfiguration.objects.current_set().filter(deprecated=True)`, which would bypass your caching. We could add similar caching to `current_set()`, or I could keep this for now, with a big TODO to remove it once there's caching in config models?

Reply:
Yeah, it's close to merge, I just dropped it when we were getting close to the LabXchange launch and didn't pick it back up again. Let me do that today, as part of this story. I think I just need some minor fixes at this point.

I did think it addressed the XBlockConfiguration issue (at least most of them), because that was the major thing I was trying to eliminate when I was doing this optimization. But we should test it out in any event. Let me get that merged, then we can measure. I'll fix up that PR right now and tag you on review.